Eliminate the `Variable` class

jagerber48 opened the pull request:
- [ ] Closes # (insert issue number)
- [ ] Executed `pre-commit run --all-files` with no errors
- [ ] The change is fully covered by automated unit tests
- [ ] Documented in docs/ as appropriate
- [ ] Added an entry to the CHANGES file

This PR:
- Eliminates the `Variable` class and eliminates the `AffineScalarFunc.derivatives` property. 
- The `AffineScalarFunc` class is renamed to `UFloat` with `AffineScalarFunc` left as a legacy name.
- The `LinearCombo` class is refactored into the `UCombo` class but the architectural role is similar: track the uncertainties of `UFloat` objects as a lazily expanded combination of objects with uncertainty
- The `UAtom` object is introduced as the fundamental element which captures uncertainty. Each `UAtom` models a random variable with zero mean and unity standard deviation/variance. Each `UAtom` has a `uuid` and all `UAtom` are mutually statistically independent. The expanded version of the `UCombo` maps `dict[UAtom, float]`.

For more detail/discussion see https://github.com/lmfit/uncertainties/discussions/251#discussioncomment-10309160

Note that as of the time opening this PR the PR is still a work in progress. Not all elements in the bullet list above are implemented and  a good handful of tests are still failing.

jagerber-qtm committed:
A first pass on linear_combo -> ucombo

jagerber-qtm committed:
revert test changes

jagerber-qtm committed:
rework UCombo, delete Variable

jagerber-qtm committed:
some test updates

jagerber-qtm committed:
Pass more tests

jagerber48 committed:
Merge branch 'master' into feature/linear_combo_refactor

jagerber48 committed:
remove large float test

jagerber48 commented on tests/test_uncertainties.py:
```diff
@@ -104,9 +105,6 @@ def test_ufloat_fromstr():
         "3.4(nan)e10": (3.4e10, float("nan")),
         # NaN value:
         "nan+/-3.14e2": (float("nan"), 314),
-        # "Double-floats"
```
These floats are too large to work with. Prior to this PR the large float value could be saved into the `AffineScalarFunc` object and displayed, but  any arithmetic with such large values would result in an `OverflowError`.  In the new framework even displaying a `UFloat` the first time requires a calculation, even if it's as simple as `sqrt(x**2)`. So in the new framework the `OverflowError` appears immediately.

jagerber48 committed:
Clean up UFloat class

jagerber48 committed:
some formatting changes

jagerber48 committed:
uncertainty instead of _linear_combo

jagerber48 committed:
variety of updates and test cleanup

jagerber48 committed:
copy tests

jagerber48 committed:
pickling tests

jagerber48 committed:
no negative std_dev in tests

jagerber48 committed:
modify test

jagerber48 committed:
more tests

jagerber48 committed:
handle some edge cases with 0 uncertainty

jagerber48 committed:
fix test_hypot

jagerber48 committed:
compare analytic and numerical derivatives on math functions

jagerber48 committed:
unumpy

jagerber48 commented:
All the tests are passing except some having to do with the uarray inverse and pseudo inverse functions. The code for this is pretty complicated. It might make sense to just rewrite the code to convert array-compatible functions to ufloat array compatible functions.

jagerber-qtm committed:
mainly code to handle inv and pinv array function

jagerber-qtm committed:
Merge remote-tracking branch 'lmfit/master' into feature/linear_combo_refactor

jagerber-qtm committed:
fix import

jagerber48 commented on tests/test_unumpy.py:
```diff
@@ -179,7 +162,7 @@ def test_wrap_array_func():
     # mat is an uncertainties.unumpy.matrix):
     def f_unc(mat, *args, **kwargs):
-        return mat.I + args[0] * kwargs["factor"]
+        return numpy.matrix(mat).I + args[0] * kwargs["factor"]
```
This function needs to cast to `numpy.matrix` internally. This is because `to_uarray_func` currently drops compatibility with `numpy.matrix`.

jagerber-qtm committed:
slim down UCombo, especially remove __hash__

jagerber-qtm committed:
Add hashes

jagerber-qtm committed:
bring back is_expanded

jagerber-qtm committed:
do not check std_dev during operations. This kills performance.

jagerber-qtm committed:
avoid empty ucombos, or ucombo/uatom weights exactly equal to zero.

We special case these away. This will roughly result in UFloats with empty uncertainty to act like normal floats. It may speed up calculations also

jagerber-qtm committed:
special case empty uncertainty in the tests. Is this something we want??

jagerber48 commented:
Prior to commit `c80dd9b` performance was terrible because on each math operation I was calculating the `std_dev` of one of the operands. But calculating `std_dev` (specifically expanding the `UCombo`) is exactly the thing which must be avoided to get the performance enhancement from the lazy evaluation. after this change and some quick benchmarking it is looking like this branch is performing similarly to `master` on the 

```
str(sum(ufloat(1, 1) for _ in range(10000)))
```
benchmark.

____________
Before merging in this PR I am going to make a new one which includes benchmarking tests. I'm not exactly sure how to code up benchmarking tests. but I'll code up some ideas and we can discuss in that PR.

jagerber48 commented:
Commits `5fc3cbd` and `6d53afb` introduces some code that special cases situations in which you have an empty uncertainty or an uncertainty with a zero weighting somewhere. These special cases were needed to ensure that something like

```
from uncertainties.umath import pow

negative = ufloat(-0.3, 0.01)
integer = ufloat(-3, 0)

pow(negative, integer)
```
gives back the same as
```
pow(negative, -3)
```
Without this special casing you end up with some kind of `nan` in the mix. Basically the special cases in those commits support the model that "a `UFloat` with zero std_dev is equivalent to a regular `float`.

Is this truly the behavior we want? Do we want these special cases?

jagerber-qtm committed:
fix broken test that wasn't running

This test wasn't running because the generator defining ufloat_args was consumed while calculating output so it couldn't be iterated over again for the for loop. Need to save the result as a list before using it.

jagerber-qtm committed:
ucombo clean up and refactor

jagerber-qtm committed:
don't calculate std_dev with kwargs inputs, cleanup f_with_affine_output a little

jagerber-qtm committed:
some correlated_values_norm cleanup

These functions need to be overhauled so for now trying to make minimal changes while keeping functionality to keep the PR as slim as possibly (though it's already pretty huge)

jagerber-qtm committed:
todo on untested, probably broken functions

jagerber-qtm committed:
codecov CI

jagerber-qtm committed:
revert previous commit (it was made to the wrong branch and had a wrong commit message)

andrewgsavage commented on uncertainties/ucombo.py:
```diff
@@ -0,0 +1,134 @@
+        return self._hash
+
+    def __lt__(self: UAtom, other: UAtom) -> bool:
+        return self.uuid < other.uuid
```
codecov gives a load of warnings for UAtom and UCombo. Would bit be worth having tests to show methods like eq, lt behave as expected?

it's also a bit suprising some of these lines (eg for UCombo's eq) don't get hit by any of the existing tests

jagerber48 replied:
Thanks for taking a look. I haven't looked at codecov yet. The story for why `UCombo`s `__eq__` doesn't get hit is worth sharing.

First off, before this PR and as of this comment, `UFloat.__eq__` does not check for equality of the nominal value and equality of the uncertainty as I would have naively implemented. Rather, it looks at the difference between `self` and `other` and checks that the resulting `UFloat` has zero nominal value and zero std_dev. Because subtraction is supported between `UFloat` and `float`, this allows `__eq__` to easily support comparison between `UFloat` and `float`. In going through the tests I've learned that, for better or for worse, `uncertainties` has definitely chosen the convention that a `UFloat` with a `std_dev` of zero should behave like a regular `float`.

I am still using this difference calculation for `UFloat.__eq__`. It would be more natural to me to have `__eq__` be something like
```
def __eq__(self, other):
    if isinstance(other, UFloat):
        return (self.nominal_value == other.nominal_value and self.uncertainty = other.uncertainty)
    else:
        if self.std_dev != 0:
            return False
        else:
            return self.nominal_value == other
```
But currently the code implementing `__eq__` is defined in `ops.py` and gets monkey patched onto `UFloat`. I can't import `UFloat` into this function because it would create circular imports between `ops.py` and `core.py`.

Anyways, I can look into disentangling this.

________________
Note though, that I tried something like the above code for a second and found out that the current version of `UCombo.__eq__` was broken because I didn't know `sorted` returned a list! Easy fix, but emphasizes the importance of testing.

Note that I made the decision that `UCombo` `__eq__` and `__hash__` use the `expanded` dictionary for comparison. Thuis means that two `UCombo`s with different `ucombo_tuple`s could compare equal. For `UFloat` I think it is clear that we want two `UFloat`s to compare based on the expanded dictionary from the uncertainty. For `UCombo` it wasn't as obvious to me. For example, if `dx` is a `UAtom` then it would be possible for there to be a nested `UCombo` storing something like
```
dx + (-dx + (dx + (-dx + (dx + (-dx + ...)))))
```
Which ultimately is equal to 0. But the former might take a longer time to compute. 

Actually I think I need to add a special case for 0 so that the presence of a `UAtom` in one `UCombo` expanded dict but not the other doesn't matter if the weight is equal to 0.

In any case, does this make sense to allow two `UCombo` to compare the same even if they have different `ucombo_tuple`? Or should that distinction be made apparent in `__eq__`  and `__hash__` (but `UFloat` will still compare based on `expanded` dict)?

jagerber-qtm committed:
fixed ucombo hash

sorted returns a (unhashable)  list, so much convert this list to a tuple before hashing.

jagerber-qtm committed:
Merge branch 'master' into feature/linear_combo_refactor

jagerber48 commented:
Ok, it is time to pick the names of the classes because I want to start updating the docs (I'm not technically blocked but would be nice to get it right earlier). Right now we have `UFloat`, UCombo` and `UAtom`. These classes are modeling the following.

For reference, suppose we have
```
x = x_0 + dx
y = y_0 + dy
z = x*y = x_0 * y_0 + (y_0 * dx + x_0 * dy)
```
`dx` and `dy` are `UAtom`s (individual random elements with zero mean and unity standard deviation/variance), `1*dx`, `1*dy`, and `y_0*dx + x_0*dy` are `UCombo`s (float-weighted linear combinations of `UAtom`s) and `x`, `y`, and `z` are `UFloat`s (nominal value + uncertainty given as a `UCombo`).

Notes on names:

- `UFloat`: I think we're all in agreement that this name is fine. `AffineScalarFunc` will remain a legacy alias that I suggest we add a `FutureWarning` too in an upcoming release (we shouldn't do that in this PR however since I don't want to get bogged down in that discussion yet. For the record, I will say there has been discussion about the name `UNumber` for something that would support more types of nominal values and uncertainties than just `floats`. E.g. `Decimal` or possibly complex numbers or something. But given `UFloat` is already in use I say it makes sense to keep it.
- `UCombo`: This class roughly replaces the old `LinearCombination` class. I don't have strong opinions on this name. `LinearCombination` would be fine, but I would like to somehow emphasize that is a linear combination **of zero-mean random variables** which is not captured by just `LinearCombination`. Other obvious options: `ULinCombo`, `ULinearCombo`, `ULinearCombination`. The same options without a `U`. Again, my opinions aren't strong here. If I don't hear otherwise I will keep `UCombo`.
- `UAtom`. I think this is the most controversial. `UAtom` kind of serves to replace the old `Variable` class, but it is much simpler than the `Variable` class and this reduction in complexity is what drives the overall improvement in the architectures (IMO). I chose the name `UAtom` because it is an atomic/irreducible/indivisible unit (this is what is meant by "atom") of uncertainty. A `UAtom` has zero mean, unity variance, and is not correlated with any other `UAtom`. This is in contrast with the `UCombo` which is a combination of many `UAtom`. However, it's been pointed out that users outside of the physical sciences may be less familiar with the term "Atom" and what it might mean here. Open to proposals for alternative names. I don't have good ideas. `UnitUncertainty`, `StdNormUnit` (standing for "standard normal unit" referring to the standard normal distribution), `RandomUnit`, `URandom`. I prefer `UAtom` to all of these but open to other proposals and discussion.

jagerber-qtm committed:
start updating docs

jagerber-qtm committed:
instance check for ABC real seems to be slow. See if removing that improves performance.

jagerber-qtm committed:
need to include int

jagerber-qtm committed:
documentation and doctest

jagerber-qtm committed:
fix index

jagerber-qtm committed:
cast default dict to dict for ucombo expanded dict. Also don't do eager checks for absence of uncertainty terms.

jagerber-qtm committed:
Add back make.bat

jagerber-qtm committed:
add doctest sphinx extension in conf.py so doctests can run

jagerber-qtm committed:
Merge branch 'feature/add_make.bat' into feature/linear_combo_refactor

jagerber-qtm committed:
UAtom repr

jagerber-qtm committed:
error_components doctest

jagerber-qtm committed:
doctests

jagerber-qtm committed:
more documentation

jagerber-qtm committed:
don't document Variable

jagerber-qtm committed:
add tuples instead of nesting for addition. Will this be faster?

jagerber48 committed:
Merge branch 'master' into feature/linear_combo_refactor

jagerber48 committed:
refactor test_power_derivatives

jagerber48 committed:
Merge branch 'master' into feature/linear_combo_refactor

jagerber48 committed:
Merge branch 'master' into feature/linear_combo_refactor

jagerber48 committed:
fixed mangled merge ordering

jagerber48 committed:
fix import

jagerber48 committed:
Merge branch 'main' into test_merge_2

jagerber48 committed:
fix some tests

jagerber48 committed:
use simpler test

jagerber48 committed:
Merge branch 'main' into test_merge_2

jagerber48 committed:
remove unused cases and generation code

jagerber48 committed:
use other case strategy

jagerber48 committed:
Merge branch 'test_merge_2' into feature/linear_combo_refactor

jagerber48 committed:
start handling zero weight differently

jagerber48 committed:
rewrite tests

jagerber48 committed:
calculated zero equality test

jagerber48 committed:
Merge branch 'main' into feature/linear_combo_refactor

jagerber48 committed:
See if uuid1 is faster

jagerber48 committed:
pickle docs

jagerber48 committed:
stop doctests for now

jagerber48 committed:
try getrandbits

jagerber48 committed:
remove checks

jagerber48 committed:
remove more checks

jagerber48 committed:
simplify hash

jagerber48 committed:
remove the bigger performance tests

jagerber48 committed:
back to hash calculation

jagerber48 committed:
Merge branch 'main' into feature/linear_combo_refactor

jagerber48 committed:
Merge branch 'main' into feature/linear_combo_refactor

jagerber48 committed:
Merge branch 'main' into feature/linear_combo_refactor

jagerber48 committed:
ValueError

jagerber48 committed:
large number changelog

jagerber48 committed:
better type annotations

jagerber48 committed:
UFloat class accepts std_dev only now for a better public API.

jagerber48 committed:
smaller uuid

jagerber48 committed:
add correlation and covariance functions

jagerber48 committed:
changelog

jagerber48 committed:
make uncertainty and UCombo stuff private

jagerber48 committed:
readme.rst

jagerber48 committed:
index.rst

jagerber48 commented:
@newville sorry, this PR originally targeted `master` but with our plan to release `4.x` from `main` this PR should merge into `main`. I've changed the PR base just now. 

I think merging this PR into `main` will be the big PR review before we release `4.0.0`. This PR is not ready for review. It still needs

(1) to address the performance regression. Before I would say it had a major performance regression. Now the regression is only ~20%. At least on the one performance benchmark we have. See #327.
(2) The documentation needs to be rewritten taking into account that new changes that is what I'm working on presently
(3) During and after the above two items I will be going over things myself and I imagine there will continue to be small changes to the source code and tests.

jagerber48 committed:
repr and str include tag

jagerber48 committed:
user_guide work

jagerber48 committed:
user guide

jagerber48 committed:
more user guide

jagerber48 committed:
user guide work

jagerber48 committed:
numpy array if possible

jagerber48 committed:
more user guide

jagerber48 committed:
more docs numpy test

jagerber48 committed:
more user guide

jagerber48 committed:
random seed

jagerber48 committed:
pickling

jagerber48 committed:
comparison

jagerber48 committed:
don't need to document `isinstance`

jagerber48 committed:
linear breakdown

jagerber48 committed:
nan docs

jagerber48 committed:
more documentation

jagerber48 committed:
ufloat as alias

jagerber48 committed:
remove special topics section

jagerber48 committed:
doc updates

jagerber48 committed:
top bar

jagerber48 committed:
pass warning

jagerber48 committed:
numpy guide

jagerber48 committed:
formatting guide

jagerber48 committed:
bring doctest back

jagerber48 committed:
formatting

jagerber48 committed:
Merge branch 'main' into feature/linear_combo_refactor

wshanks commented on uncertainties/ucombo.py:
```diff
@@ -0,0 +1,141 @@
+
+    def __init__(self: UAtom, tag: Optional[str] = None):
+        self.tag = tag
+        self.uuid: int = random.getrandbits(64)
```
I just wanted to note how Python ints work in practice. They are variable in size. There is a 28 byte overhead for all `int`s. Beyond that, every 30 bits in size adds 4 more bytes. So getting 64 bits will give 36 byte objects (except when you randomly get a number smaller than 2**59). Maybe it would be better to do `getrandbits(60)` here so the overhead per `UAtom` is 32 bytes instead of 36 if 60 bits seems like enough randomness. I wondered if `randbytes()` would be more efficient but `bytes` has an overhead of 34 bytes instead of the 28 for int. bytes scales linearly in the number of bytes stored but remains less efficient than int at these small sizes (8 bytes / 64 bits would be 42 bytes total). A float would take only 24 bytes but I worry about populating those bytes in an unbiased way.
